### PR TITLE
[X86] Respect code models more when determining if a global reference can fit in 32 bits

### PR DESCRIPTION
**Pull request by aeubanks**
For non-GlobalValue references, the small and medium code models can use 32 bit constants.

For GlobalValue references, use TargetMachine::isLargeGlobalObject().
Look through aliases for determining if a GlobalValue is small or large.
Even the large code model can reference small objects with 32 bit constants as long as we're in no-pic mode, or if the reference is offset from the GOT.

### CODE REVIEW

File: llvm/lib/Target/X86/X86Subtarget.cpp [hunk 1]
```diff
@@ -91,9 +91,8 @@ X86Subtarget::classifyLocalReference(const GlobalValue *GV) const {
       if (CM == CodeModel::Large)
```
Comment (rnk) [inline llvm/lib/Target/X86/X86Subtarget.cpp:91]: Is this condition correct? It seems like it won't handle globals explicitly marked with `code_model "small"` correctly.

File: llvm/lib/Target/X86/X86Subtarget.cpp [hunk 2]
```diff
@@ -91,9 +91,8 @@ X86Subtarget::classifyLocalReference(const GlobalValue *GV) const {
...
         return X86II::MO_GOTOFF;
       // Large objects use GOTOFF, otherwise use RIP-rel access.
-      if (auto *GO = dyn_cast_or_null<GlobalObject>(GV))
-        return TM.isLargeGlobalObject(GO) ? X86II::MO_GOTOFF
-                                          : X86II::MO_NO_FLAG;
+      if (GV)
+        return TM.isLargeGlobalValue(GV) ? X86II::MO_GOTOFF : X86II::MO_NO_FLAG;
       // For non-GlobalObjects, the small and medium code models treat them as
```
Comment (rnk) [inline llvm/lib/Target/X86/X86Subtarget.cpp:96]: This comment seems stale. What does it mean if GV is null? Does that represent a reference to a local label, something like a constant pool?

File: llvm/test/CodeGen/X86/code-model-elf.ll [hunk 3]
```diff
@@ -96,12 +97,67 @@ define dso_local ptr @lea_static_data() #0 {
...
+; MEDIUM-PIC-NEXT:    leaq _GLOBAL_OFFSET_TABLE_(%rip), %rcx
+; MEDIUM-PIC-NEXT:    movabsq $static_data_alias@GOTOFF, %rax
+; MEDIUM-PIC-NEXT:    addq %rcx, %rax
+; MEDIUM-PIC-NEXT:    retq
+;
+; LARGE-PIC-LABEL: lea_static_data_alias:
+; LARGE-PIC:       # %bb.0:
+; LARGE-PIC-NEXT:  .L1$pb:
+; LARGE-PIC-NEXT:    leaq .L1$pb(%rip), %rax
+; LARGE-PIC-NEXT:    movabsq $_GLOBAL_OFFSET_TABLE_-.L1$pb, %rcx
```
Comment (rnk) [inline llvm/test/CodeGen/X86/code-model-elf.ll:143]: The large PIC code model is awful. Is there really no way to express `$static_data_alias - .L1$pb` with a static relocation? That would avoid the need for the second add and i64 materialization.

File: llvm/test/CodeGen/X86/code-model-elf.ll [hunk 4]
```diff
@@ -1072,7 +1127,7 @@ define dso_local float @load_constant_pool(float %x) #0 {
 ;
 ; MEDIUM-STATIC-LABEL: load_constant_pool:
 ; MEDIUM-STATIC:       # %bb.0:
-; MEDIUM-STATIC-NEXT:    movabsq ${{\.?LCPI[0-9]+_[0-9]+}}, %rax
+; MEDIUM-STATIC-NEXT:    movl ${{\.?LCPI[0-9]+_[0-9]+}}, %eax
```
Comment (rnk) [inline llvm/test/CodeGen/X86/code-model-elf.ll:1130]: These constant pool changes look correct.

File: llvm/test/CodeGen/X86/code-model-elf.ll (hunk 3 above)
Comment (aeubanks) [reply llvm/test/CodeGen/X86/code-model-elf.ll:143]: the problem is relocations only allow one symbol to relocate against, we can't just diff two symbols. so we have to take advantage of the other implicit operands in the relocations, e.g. the PC/GOT/PLT, for PIC. for example, using the GOT as an anchor requires a 64-bit relocation to get the address of the GOT. the only anchor that x86-64 static relocations provide that can be assumed to be close to the instruction is the PC, so we'd probably need to use R_X86_64_PC64 (seems like this is a rare relocation type). perhaps something like the following

```
  leaq .L(%rip), %rax
.L
  movabsq foo-%rip, %rbx
  addq %rbx, %rax
```

but now you need a label for every reference to a global. I'm not sure how much this impacts optimizations. (do optimizations take care to not move around labels if their relative placement with instructions matter?)

maybe this is worth experimenting with at some point

Comment (rnk) [reply llvm/test/CodeGen/X86/code-model-elf.ll:143]: Makes sense. I suppose the GOT materialization only happens once per function anyway, so each symbol reference really only requires one movabsq & add. So, it's not really worth further exploration.

**Review by rnk — APPROVED** [review]
Thanks, looks good!

**Comment by ormris** [comment]
+1 We're also seeing this failure in our internal CI.

**Comment by aeubanks** [comment]
reverted, taking a look

**Pull request merged**